crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -5475,6 +5483,99 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         }
     }
 
+    /// Infer the implicit local definition `x = <module 'thispackage.x'>` that
+    /// `from .x.y import z` can introduce in an `__init__.py(i)`.
+    ///
+    /// For the definition `z`, see [`TypeInferenceBuilder::infer_import_from_definition`].
+    fn infer_import_from_implicit_definition(
+        &mut self,
+        import_from: &ast::StmtImportFrom,
+        submodule: &Name,
+        definition: Definition<'db>,
+    ) {
+        // Although the *actual* runtime semantic of this kind of statement is to
+        // introduce a variable in the global scope of this module, we want to
+        // encourage users to write code that doesn't have dependence on execution-order.
+        //
+        // By introducing it as a local variable in the scope the import occurs in,
+        // we effectively require the developer to either do the import at the start of
+        // the file where it belongs, or to repeat the import in every function that
+        // wants to use it, which "definitely" works.
+        //
+        // (It doesn't actually "definitely" work because only the first import of `thispackage.x`
+        // will ever set `x`, and any subsequent overwrites of it will permanently clobber it.
+        // Also, a local variable `x` in a function should always shadow the submodule because
+        // the submodule is defined at file-scope. However, both of these issues are much more
+        // narrow, so this approach seems to work well in practice!)
```

Comment:
As discussed in person, I don't think accepting these two bugs is the right tradeoff, vs just not modeling submodule-attribute side-effects at all for imports inside functions. Ok with landing this as-is, as long as there is an upcoming subsequent PR that adjusts this to simply not create these "implicit" definitions at all inside functions.